compiler/rustc_hir_analysis/src/delegation.rs

```diff
@@ -76,24 +80,60 @@ fn fn_kind<'tcx>(tcx: TyCtxt<'tcx>, def_id: DefId) -> FnKind {
     }
 }
 
+/// Given the current context(caller and callee `FnKind`), it specifies
+/// the policy of predicates and generic parameters inheritance.
+#[derive(Clone, Copy, Debug, PartialEq)]
+enum InheritanceKind {
+    /// Copying all predicates and parameters, including those of the parent
+    /// container.
+    ///
+    /// Boolean value defines whether the `Self` parameter or `Self: Trait`
+    /// predicate are copied. It's always equal to `true` except when
+    /// delegating from a free function to a trait method.
+    ///
+    /// FIXME(fn_delegation): Eagerly copying often leads to type inference
+    /// errors. Support providing generic arguments or restrict use sites.
+    Eagerly(bool),
```

Comment:
"Eager" is not a fitting word here, maybe just call it "with parent"/"without parent".